File: app/src/main/java/com/manimarank/spell4wiki/activities/Spell4WordListActivity.java
package com.manimarank.spell4wiki.activities;

import android.app.Activity;
import android.app.AlertDialog;
import android.content.Intent;
import android.database.Cursor;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.provider.MediaStore;
import android.text.TextUtils;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import androidx.annotation.RequiresApi;
import androidx.interpolator.view.animation.FastOutSlowInInterpolator;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.manimarank.spell4wiki.R;
import com.manimarank.spell4wiki.activities.base.BaseActivity;
import com.manimarank.spell4wiki.adapters.EndlessRecyclerAdapter;
import com.manimarank.spell4wiki.databases.DBHelper;
import com.manimarank.spell4wiki.databases.dao.WordsHaveAudioDao;
import com.manimarank.spell4wiki.databases.entities.WordsHaveAudio;
import com.manimarank.spell4wiki.fragments.LanguageSelectionFragment;
import com.manimarank.spell4wiki.listerners.OnLanguageSelectionListener;
import com.manimarank.spell4wiki.utils.ExtensionsKt;
import com.manimarank.spell4wiki.utils.GeneralUtils;
import com.manimarank.spell4wiki.utils.PrefManager;
import com.manimarank.spell4wiki.utils.RealPathUtil;
import com.manimarank.spell4wiki.utils.ShowCasePref;
import com.manimarank.spell4wiki.utils.SnackBarUtils;
import com.manimarank.spell4wiki.utils.constants.AppConstants;
import com.manimarank.spell4wiki.views.EndlessRecyclerView;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;

import uk.co.samuelwall.materialtaptargetprompt.MaterialTapTargetPrompt;
import uk.co.samuelwall.materialtaptargetprompt.MaterialTapTargetSequence;
import uk.co.samuelwall.materialtaptargetprompt.extras.focals.RectanglePromptFocal;

import static com.manimarank.spell4wiki.utils.constants.EnumTypeDef.ListMode.SPELL_4_WORD_LIST;


public class Spell4WordListActivity extends BaseActivity {


    private EditText editFile;
    private TextView txtFileInfo;
    private View layoutEdit, layoutSelect, layoutEmpty;

    private EndlessRecyclerView recyclerView;
    private EndlessRecyclerAdapter adapter;
    private String languageCode = "";
    private WordsHaveAudioDao wordsHaveAudioDao;

    private PrefManager pref;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_spell_4_wordlist);

        pref = new PrefManager(getApplicationContext());
        languageCode = pref.getLanguageCodeSpell4WordList();

        initUI();
    }

    private void initUI() {

        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
            getSupportActionBar().setTitle(getString(R.string.spell4wordlist));
        }

        wordsHaveAudioDao = DBHelper.getInstance(getApplicationContext()).getAppDatabase().getWordsHaveAudioDao();

        Button btnSelectFile = findViewById(R.id.btnSelectFile);
        Button btnDirectContent = findViewById(R.id.btnDirectContent);
        Button btnDone = findViewById(R.id.btnDone);
        editFile = findViewById(R.id.editFile);
        // Remove styles after paste/added content
        ExtensionsKt.removeStyleAfterPaste(editFile);
        txtFileInfo = findViewById(R.id.txtFileInfo);
        recyclerView = findViewById(R.id.recyclerView);
        layoutSelect = findViewById(R.id.layoutSelect);
        layoutEdit = findViewById(R.id.layoutEdit);
        layoutEmpty = findViewById(R.id.layoutEmpty);


        btnSelectFile.setOnClickListener(v -> {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
                editDocument();
            }
        });

        btnDirectContent.setOnClickListener(v -> showDirectContentAlignMode());

        btnDone.setOnClickListener(v -> {
            GeneralUtils.hideKeyboard(Spell4WordListActivity.this);
            if (!TextUtils.isEmpty(editFile.getText())) {
                List<String> items = getWordListFromString(editFile.getText().toString());
                showWordsInRecordMode(items);
            } else
                SnackBarUtils.INSTANCE.showLong(editFile, getString(R.string.provide_valid_content));

        });


        layoutSelect.setVisibility(View.VISIBLE);
        layoutEdit.setVisibility(View.GONE);
        recyclerView.setVisibility(View.GONE);
        layoutEmpty.setVisibility(View.GONE);

    }


    /**
     * Open a file for writing and append some text to it.
     */
    @RequiresApi(api = Build.VERSION_CODES.KITKAT)
    private void editDocument() {
        // ACTION_OPEN_DOCUMENT is the intent to choose a file via the system's
        // file browser.
        Intent intent = new Intent(Intent.ACTION_OPEN_DOCUMENT);

        // Filter to only show results that can be "opened", such as a
        // file (as opposed to a list of contacts or timezones).
        intent.addCategory(Intent.CATEGORY_OPENABLE);

        // Filter to show only text files.
        intent.setType("text/plain");

        startActivityForResult(intent, AppConstants.RC_EDIT_REQUEST_CODE);
    }

    public void updateList(String word) {
        if(!isDestroyed() && !isFinishing()) {
            if (adapter != null) {
                wordsHaveAudioDao.insert(new WordsHaveAudio(word, languageCode));
                adapter.addWordInWordsHaveAudioList(word);
                adapter.remove(word);
                if (adapter.getItemCount() == 0)
                    showEmptyView();
            }
        }
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode,
                                 Intent data) {

        // The ACTION_OPEN_DOCUMENT intent was sent with the request code
        // READ_REQUEST_CODE. If the request code seen here doesn't match, it's the
        // response to some other intent, and the code below shouldn't run at all.

        super.onActivityResult(requestCode, resultCode, data);
        if(!isDestroyed() && !isFinishing()) {
            if (requestCode == AppConstants.RC_EDIT_REQUEST_CODE && resultCode == Activity.RESULT_OK) {
                // The document selected by the user won't be returned in the intent.
                // Instead, a URI to that document will be contained in the return intent
                // provided to this method as a parameter.
                // Pull that URI using resultData.getData().
                Uri uri;
                if (data != null) {
                    uri = data.getData();
                    assert uri != null;
                    File file = new File(RealPathUtil.getRealPath(getApplicationContext(), uri));
                    openFileInAlignMode(file.getAbsolutePath());
                }
            }

            if (requestCode == AppConstants.RC_UPLOAD_DIALOG) {
                if (data != null && data.hasExtra(AppConstants.WORD)) {
                    if (adapter != null) {
                        adapter.addWordInWordsHaveAudioList(data.getStringExtra(AppConstants.WORD));
                        adapter.remove(data.getStringExtra(AppConstants.WORD));
                        if (adapter.getItemCount() == 0)
                            showEmptyView();
                    }
                }
            }
        }
    }

    private void openFileInAlignMode(String filePath) {

        layoutSelect.setVisibility(View.GONE);
        layoutEdit.setVisibility(View.VISIBLE);
        recyclerView.setVisibility(View.GONE);
        layoutEmpty.setVisibility(View.GONE);

        txtFileInfo.setText(getString(R.string.hint_select_file_next));
        editFile.setText(getContentFromFile(filePath));
    }

    private void showDirectContentAlignMode() {
        layoutSelect.setVisibility(View.GONE);
        layoutEdit.setVisibility(View.VISIBLE);
        recyclerView.setVisibility(View.GONE);
        layoutEmpty.setVisibility(View.GONE);

        txtFileInfo.setText(getString(R.string.hint_direct_copy_next));
        editFile.setText("");
    }

    private void showWordsInRecordMode(List<String> items) {

        List<String> wordsAlreadyHaveAudio = wordsHaveAudioDao.getWordsAlreadyHaveAudioByLanguage(languageCode);
        items.removeAll(wordsAlreadyHaveAudio);
        if (items.size() > 0) {
            layoutSelect.setVisibility(View.GONE);
            layoutEdit.setVisibility(View.GONE);
            recyclerView.setVisibility(View.VISIBLE);
            layoutEmpty.setVisibility(View.GONE);

            recyclerView.setHasFixedSize(true);
            LinearLayoutManager layoutManager = new LinearLayoutManager(this);
            recyclerView.setLayoutManager(layoutManager);

            adapter = new EndlessRecyclerAdapter(this, items, SPELL_4_WORD_LIST);
            recyclerView.setAdapter(adapter, layoutManager);
            adapter.setWordsHaveAudioList(wordsHaveAudioDao.getWordsAlreadyHaveAudioByLanguage(languageCode));

            if (ShowCasePref.INSTANCE.isNotShowed(ShowCasePref.LIST_ITEM_SPELL_4_WIKI))
                new Handler().post(this::callShowCaseUI);
        } else {
            showEmptyView();
        }
    }

    private void showEmptyView() {
        layoutSelect.setVisibility(View.GONE);
        layoutEdit.setVisibility(View.GONE);
        recyclerView.setVisibility(View.GONE);
        layoutEmpty.setVisibility(View.VISIBLE);
    }

    private List<String> getWordListFromString(String data) {

        List<String> list = new ArrayList<>();
        try {
            if (data != null && data.length() > 0) {
                String[] l = data.split("\n");

                if (l.length > 0) {
                    for (String s : l) {
                        if (s != null) {
                            String word = s.trim();
                            if (word.length() > 0) {
                                list.add(word);
                            }
                        }
                    }
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }

        return list;
    }

    private String getRealPathFromURI(Uri contentURI) {
        String result = null;
        Cursor cursor = null;//ww  w.j  a  va2  s. c om
        try {
            String[] proj = {MediaStore.Images.Media.DATA};
            cursor = getContentResolver().query(contentURI, proj,
                    null, null, null);
            int column_index = cursor != null ? cursor.getColumnIndexOrThrow(MediaStore.Images.Media.DATA) : 0;
            if (cursor != null && cursor.moveToFirst())
                result = cursor.getString(column_index);
        } finally {
            if (cursor != null) {
                cursor.close();
            }
        }

        return result;
    }

    private String getContentFromFile(String fileName) {
        String data = "";
        try {
            FileInputStream inputStream = new FileInputStream(fileName);
            InputStreamReader inputStreamReader = new InputStreamReader(inputStream);
            BufferedReader reader = new BufferedReader(inputStreamReader);

            String line;

            // Reading line by line from the
            // file until a null is returned
            StringBuilder stringBuilder = new StringBuilder();

            while ((line = reader.readLine()) != null) {
                stringBuilder.append(line).append("\n");
            }

            inputStream.close();
            data = stringBuilder.toString().trim();

            /*while ((line = reader.readLine()) != null) {
                String s = line.trim();
                if (s.length() > 0)
                    list.add(line);
            }*/

        } catch (Exception e) {
            e.printStackTrace();
        }


        return data;
    }

    private List<String> readFromFile(String fileName) {
        List<String> list = new ArrayList<>();
        try {
            FileInputStream inputStream = new FileInputStream(fileName);
            InputStreamReader inputStreamReader = new InputStreamReader(inputStream);
            BufferedReader reader = new BufferedReader(inputStreamReader);

            String line;

            // Reading line by line from the
            // file until a null is returned
            while ((line = reader.readLine()) != null) {
                String s = line.trim();
                if (s.length() > 0)
                    list.add(line);
            }

        } catch (Exception e) {
            e.printStackTrace();
        }

        return list;
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.spell4wiki_view_menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem menuItem) {
        if (menuItem.getItemId() == android.R.id.home) {
            callBackPress();
        }
        return (super.onOptionsItemSelected(menuItem));
    }

    private void loadLanguages() {
        OnLanguageSelectionListener callback = langCode -> {
            if (!languageCode.equals(langCode)) {
                languageCode = langCode;
                invalidateOptionsMenu();
                if (recyclerView.getVisibility() == View.VISIBLE || layoutEmpty.getVisibility() == View.VISIBLE) {
                    if (!TextUtils.isEmpty(editFile.getText())) {
                        List<String> items = getWordListFromString(editFile.getText().toString());
                        showWordsInRecordMode(items);
                    }
                }
            }
        };
        LanguageSelectionFragment languageSelectionFragment = new LanguageSelectionFragment(this);
        languageSelectionFragment.init(callback, SPELL_4_WORD_LIST);
        languageSelectionFragment.show(getSupportFragmentManager());
    }

    private void setupLanguageSelectorMenuItem(Menu menu) {
        MenuItem item = menu.findItem(R.id.menu_lang_selector);
        item.setVisible(true);
        View rootView = item.getActionView();
        TextView selectedLang = rootView.findViewById(R.id.txtSelectedLanguage);
        selectedLang.setText(this.languageCode.toUpperCase());
        rootView.setOnClickListener(v -> loadLanguages());
    }

    @Override
    public boolean onPrepareOptionsMenu(Menu menu) {
        setupLanguageSelectorMenuItem(menu);
        return super.onPrepareOptionsMenu(menu);
    }

    @Override
    public void onBackPressed() {
        callBackPress();
    }

    private void callBackPress() {
        if (recyclerView.getVisibility() == View.VISIBLE || layoutEmpty.getVisibility() == View.VISIBLE) {
            layoutEdit.setVisibility(View.VISIBLE);
            recyclerView.setVisibility(View.GONE);
            layoutEmpty.setVisibility(View.GONE);
        } else if (layoutEdit.getVisibility() == View.VISIBLE && pref.getAbortAlertStatus()) {
            if (!TextUtils.isEmpty(editFile.getText())) {
                AlertDialog.Builder builder = new AlertDialog.Builder(this);
                builder.setTitle(R.string.confirmation);
                builder.setMessage(R.string.confirm_to_back);
                builder.setCancelable(false);
                builder.setPositiveButton(getString(R.string.yes), (dialog, which) -> {
                    layoutSelect.setVisibility(View.VISIBLE);
                    layoutEdit.setVisibility(View.GONE);
                });
                builder.setNegativeButton(getString(R.string.cancel), (dialog, which) -> {
                });
                AlertDialog dialog = builder.create();
                dialog.show();
            } else {
                layoutSelect.setVisibility(View.VISIBLE);
                layoutEdit.setVisibility(View.GONE);
            }
        } else
            super.onBackPressed();
    }

    private void callShowCaseUI() {
        if (!isFinishing() && !isDestroyed()) {
            if (ShowCasePref.INSTANCE.isNotShowed(ShowCasePref.LIST_ITEM_SPELL_4_WIKI) && recyclerView != null && recyclerView.getVisibility() == View.VISIBLE && recyclerView.getChildAt(0) != null) {
                MaterialTapTargetSequence sequence = new MaterialTapTargetSequence().setSequenceCompleteListener(() -> ShowCasePref.INSTANCE.showed(ShowCasePref.LIST_ITEM_SPELL_4_WIKI));
                sequence.addPrompt(getPromptBuilder()
                        .setTarget(recyclerView.getChildAt(0))
                        .setPrimaryText(R.string.sc_t_spell4wiki_list_item)
                        .setSecondaryText(R.string.sc_d_spell4wiki_list_item));
                sequence.show();
            }
        }
    }

    private MaterialTapTargetPrompt.Builder getPromptBuilder() {
        return new MaterialTapTargetPrompt.Builder(Spell4WordListActivity.this)
                .setPromptFocal(new RectanglePromptFocal())
                .setAnimationInterpolator(new FastOutSlowInInterpolator())
                .setFocalPadding(R.dimen.show_case_focal_padding);
    }
}

File: app/src/main/java/com/manimarank/spell4wiki/ui/spell4wiktionary/Spell4WordListActivity.kt
package com.manimarank.spell4wiki.ui.spell4wiktionary

import android.app.Dialog
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.os.Handler
import android.text.TextUtils
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.Window
import android.widget.Button
import android.widget.TextView
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.widget.Toolbar
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import com.manimarank.spell4wiki.R
import com.manimarank.spell4wiki.data.db.DBHelper
import com.manimarank.spell4wiki.data.db.dao.WikiLangDao
import com.manimarank.spell4wiki.data.db.dao.WordsHaveAudioDao
import com.manimarank.spell4wiki.data.db.entities.WordsHaveAudio
import com.manimarank.spell4wiki.data.prefs.PrefManager
import com.manimarank.spell4wiki.data.prefs.ShowCasePref
import com.manimarank.spell4wiki.data.prefs.ShowCasePref.isNotShowed
import com.manimarank.spell4wiki.data.prefs.ShowCasePref.showed
import com.manimarank.spell4wiki.ui.common.BaseActivity
import com.manimarank.spell4wiki.ui.dialogs.CommonDialog.openRunFilterInfoDialog
import com.manimarank.spell4wiki.ui.dialogs.showConfirmBackDialog
import com.manimarank.spell4wiki.ui.languageselector.LanguageSelectionFragment
import com.manimarank.spell4wiki.ui.listerners.OnLanguageSelectionListener
import com.manimarank.spell4wiki.utils.EdgeToEdgeUtils.setupEdgeToEdgeWithToolbar
import com.manimarank.spell4wiki.utils.GeneralUtils
import com.manimarank.spell4wiki.utils.GeneralUtils.getPromptBuilder
import com.manimarank.spell4wiki.utils.GeneralUtils.hideKeyboard
import com.manimarank.spell4wiki.utils.SnackBarUtils.showLong
import com.manimarank.spell4wiki.utils.constants.AppConstants
import com.manimarank.spell4wiki.utils.constants.ListMode
import com.manimarank.spell4wiki.utils.makeGone
import com.manimarank.spell4wiki.utils.makeVisible
import com.manimarank.spell4wiki.databinding.ActivitySpell4WordlistBinding
import uk.co.samuelwall.materialtaptargetprompt.MaterialTapTargetSequence
import java.io.BufferedReader
import java.io.FileInputStream
import java.io.IOException
import java.io.InputStreamReader
import java.util.Locale


class Spell4WordListActivity : BaseActivity() {

    private lateinit var binding: ActivitySpell4WordlistBinding
    private var adapter: EndlessRecyclerAdapter? = null
    private var languageCode: String? = ""
    private var wikiLangDao: WikiLangDao? = null
    private var wordsHaveAudioDao: WordsHaveAudioDao? = null
    private lateinit var pref: PrefManager
    private lateinit var viewModel: MainViewModel
    private val filterRemovedWords = ArrayList<String>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivitySpell4WordlistBinding.inflate(layoutInflater)
        setContentView(binding.root)
        pref = PrefManager(applicationContext)
        languageCode = pref.languageCodeSpell4WikiAll
        initUI()
    }

    private fun initUI() {
        wikiLangDao = DBHelper.getInstance(applicationContext).appDatabase.wikiLangDao
        wordsHaveAudioDao = DBHelper.getInstance(applicationContext).appDatabase.wordsHaveAudioDao

        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        val wikiLang = wikiLangDao?.getWikiLanguageWithCode(languageCode)

        // Setup proper status bar handling
        setupEdgeToEdgeWithToolbar(
            rootView = binding.root,
            toolbar = toolbar
        )

        setSupportActionBar(toolbar)
        supportActionBar?.title = getString(R.string.spell4wordlist)
        supportActionBar?.subtitle = GeneralUtils.getLanguageInfo(applicationContext, wikiLang)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        binding.btnSelectFile.setOnClickListener {
            openTextFilePicker()
        }
        binding.btnDirectContent.setOnClickListener { showDirectContentAlignMode() }
        binding.btnDone.setOnClickListener {
            hideKeyboard(this@Spell4WordListActivity)
            if (!TextUtils.isEmpty(binding.editFile.text)) {
                val items = getWordListFromString(binding.editFile.text.toString())
                showWordsInRecordMode(items)
            } else showLong(binding.editFile, getString(R.string.provide_valid_content))
        }
        binding.layoutSelect.makeVisible()
        binding.layoutEdit.makeGone()
        binding.layoutList.makeGone()
        binding.root.findViewById<View>(R.id.layoutEmpty).makeGone()

        viewModel = ViewModelProvider(this)[MainViewModel::class.java]
        setupFilterWordOption()
    }

    /**
     * Open a text file to read the words
     */
    private val openTextFileLauncher = registerForActivityResult( ActivityResultContracts.OpenDocument(), this::handleSelectedTextFile)

    // Handle the result of file selection
    private fun handleSelectedTextFile(uri: Uri?) {
        if (uri != null) {
            readTextFileFromUri(uri)
        }
    }

    // Read content of the selected text file
    private fun readTextFileFromUri(uri: Uri) {
        try {
            contentResolver.openInputStream(uri).use { inputStream ->
                BufferedReader(InputStreamReader(inputStream)).use { reader ->
                    val fileContent = java.lang.StringBuilder()
                    var line: String?
                    while ((reader.readLine().also { line = it }) != null) {
                        fileContent.append(line).append("\n")
                    }
                    openFileContentInAlignMode(fileContent.toString())
                }
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    // Launch SAF to pick a text file
    private fun openTextFilePicker() {
        openTextFileLauncher.launch(arrayOf("text/plain")) // MIME type for text files
    }

    fun updateList(word: String?) {
        if (!isDestroyed && !isFinishing) {
            if (adapter != null) {
                wordsHaveAudioDao?.insert(WordsHaveAudio(word, languageCode))
                adapter?.addWordInWordsHaveAudioList(word)
                adapter?.remove(word)
                if (adapter?.itemCount == 0) showEmptyView()
            }
        }
    }

    private fun setupFilterWordOption() {
        val dialog = Dialog(this, R.style.RecordAudioDialogTheme)
        dialog.requestWindowFeature(Window.FEATURE_NO_TITLE)
        dialog.setContentView(R.layout.loading_file_availability)
        val txtInfo = dialog.findViewById<TextView>(R.id.txtFileName)
        val txtProgress = dialog.findViewById<TextView>(R.id.txtProgress)
        val btnCancel = dialog.findViewById<Button>(R.id.btnCancel)
        txtProgress.makeVisible()
        txtInfo.text = getFilterText("")
        dialog.setCancelable(false)

        var itemList: List<String> = ArrayList()
        filterRemovedWords.clear()

        viewModel.progressForFilter.observe(this) { index ->
            // val progress = ((index+1) * 100) / itemList.size
            txtProgress.text = ("${index + 1}/${itemList.size}")
            txtInfo.text = getFilterText(itemList[index])
        }

        viewModel.wordAlreadyHaveAudio.observe(this) { word ->
            updateList(word)
        }

        viewModel.wordsWithoutAudioList.observe(this) { list ->
            val diff = itemList.size - list.size
            showLong(
                binding.recyclerView,
                if (diff > 0) getString(
                    R.string.words_filter_success,
                    diff
                ) else getString(R.string.no_words_filtered)
            )
            filterRemovedWords.addAll(list)
            dialog.dismiss()
        }

        viewModel.filterCancelled.observe(this) { cancelled ->
            if (cancelled) {
                showLong(binding.recyclerView, getString(R.string.filter_cancelled))
                dialog.dismiss()
            }
        }

        btnCancel.setOnClickListener {
            viewModel.cancelFilter()
        }

        binding.root.findViewById<View>(R.id.btnRunFilter).setOnClickListener {
            val runFilterNoOfWordsCheckCount = pref.runFilterNumberOfWordsToCheck ?: AppConstants.RUN_FILTER_NO_OF_WORDS_CHECK_COUNT
            itemList = adapter?.getList()?.filter { filterRemovedWords.contains(it).not() }?.take(runFilterNoOfWordsCheckCount) ?: listOf()
            if (itemList.isNotEmpty() && languageCode != null) {
                txtProgress.text = ("0/${itemList.size}")
                dialog.show()
                viewModel.checkWordsAvailability(itemList, languageCode!!, runFilterNoOfWordsCheckCount)
            } else
                showLong(binding.recyclerView, getString(R.string.no_words_scroll_to_get_new_words))
        }

        binding.root.findViewById<View>(R.id.btnRunFilterInfo).setOnClickListener { this.openRunFilterInfoDialog() }
    }

    private fun getFilterText(word: String?): String {
        val fileName = "$languageCode-$word.ogg"
        return String.format(getString(R.string.checking_file_availability), fileName)
    }

    public override fun onActivityResult(
        requestCode: Int, resultCode: Int,
        data: Intent?
    ) {

        // The ACTION_OPEN_DOCUMENT intent was sent with the request code
        // READ_REQUEST_CODE. If the request code seen here doesn't match, it's the
        // response to some other intent, and the code below shouldn't run at all.
        super.onActivityResult(requestCode, resultCode, data)
        if (!isDestroyed && !isFinishing) {
            if (requestCode == AppConstants.RC_UPLOAD_DIALOG) {
                if (data != null && data.hasExtra(AppConstants.WORD)) {
                    adapter?.addWordInWordsHaveAudioList(data.getStringExtra(AppConstants.WORD))
                    adapter?.remove(data.getStringExtra(AppConstants.WORD))
                    if (adapter?.itemCount == 0)
                        showEmptyView()
                }
            }
        }
    }

    private fun openFileContentInAlignMode(fileContent: String) {
        binding.layoutSelect.makeGone()
        binding.layoutEdit.makeVisible()
        binding.layoutList.makeGone()
        binding.root.findViewById<View>(R.id.layoutEmpty).makeGone()
        binding.txtFileInfo.text = getString(R.string.hint_select_file_next)
        binding.editFile.setText(fileContent)
    }

    private fun showDirectContentAlignMode() {
        binding.layoutSelect.makeGone()
        binding.layoutEdit.makeVisible()
        binding.layoutList.makeGone()
        binding.root.findViewById<View>(R.id.layoutEmpty).makeGone()
        binding.txtFileInfo.text = getString(R.string.hint_direct_copy_next)
        binding.editFile.setText("")
    }

    private fun showWordsInRecordMode(items: MutableList<String>) {
        val wordsAlreadyHaveAudio = wordsHaveAudioDao?.getWordsAlreadyHaveAudioByLanguage(languageCode)
        wordsAlreadyHaveAudio?.let {
            items.removeAll(wordsAlreadyHaveAudio)
        }
        if (items.size > 0) {
            binding.layoutSelect.makeGone()
            binding.layoutEdit.makeGone()
            binding.layoutList.makeVisible()
            binding.root.findViewById<View>(R.id.layoutEmpty).makeGone()
            binding.recyclerView.setHasFixedSize(true)
            val layoutManager = LinearLayoutManager(this)
            binding.recyclerView.layoutManager = layoutManager
            adapter = EndlessRecyclerAdapter(this, items, ListMode.SPELL_4_WORD_LIST)
            binding.recyclerView.setAdapter(adapter, layoutManager)
            adapter?.setWordsHaveAudioList(wordsHaveAudioDao?.getWordsAlreadyHaveAudioByLanguage(languageCode)?.toMutableList())
            if (isNotShowed(ShowCasePref.LIST_ITEM_SPELL_4_WIKI))
                Handler().post { callShowCaseUI() }
        } else {
            showEmptyView()
        }
    }

    private fun showEmptyView() {
        binding.layoutSelect.makeGone()
        binding.layoutEdit.makeGone()
        binding.layoutList.makeGone()
        binding.root.findViewById<View>(R.id.layoutEmpty).makeVisible()
    }

    private fun getWordListFromString(data: String?): MutableList<String> {
        val list: MutableList<String> = ArrayList()
        try {
            if (data != null && data.isNotEmpty()) {
                val l = data.split("\n").toTypedArray()
                if (l.isNotEmpty()) {
                    for (s in l) {
                        val word = s.trim { it <= ' ' }
                        if (word.isNotEmpty() && !list.contains(word)) {
                            list.add(word)
                        }
                    }
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return list
    }

    private fun getContentFromFile(fileName: String): String {
        var data = ""
        try {
            val inputStream = FileInputStream(fileName)
            val inputStreamReader = InputStreamReader(inputStream)
            val reader = BufferedReader(inputStreamReader)
            var line: String?

            // Reading line by line from the
            // file until a null is returned
            val stringBuilder = StringBuilder()
            while (reader.readLine().also { line = it } != null) {
                stringBuilder.append(line).append("\n")
            }
            inputStream.close()
            data = stringBuilder.toString().trim { it <= ' ' }

        } catch (e: Exception) {
            e.printStackTrace()
        }
        return data
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.spell4wiki_view_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(menuItem: MenuItem): Boolean {
        if (menuItem.itemId == android.R.id.home) {
            callBackPress()
        }
        return super.onOptionsItemSelected(menuItem)
    }

    private fun loadLanguages() {
        val callback = object : OnLanguageSelectionListener {
            override fun onCallBackListener(langCode: String?) {
                if (languageCode != langCode) {
                    languageCode = langCode
                    supportActionBar?.subtitle = GeneralUtils.getLanguageInfo(applicationContext, wikiLangDao?.getWikiLanguageWithCode(langCode))
                    invalidateOptionsMenu()
                    if (binding.layoutList.visibility == View.VISIBLE || binding.root.findViewById<View>(R.id.layoutEmpty).visibility == View.VISIBLE) {
                        if (!TextUtils.isEmpty(binding.editFile.text)) {
                            val items = getWordListFromString(
                                binding.editFile.text.toString()
                            )
                            showWordsInRecordMode(items)
                        }
                    }
                }
            }
        }
        val languageSelectionFragment = LanguageSelectionFragment(this)
        languageSelectionFragment.init(callback, ListMode.SPELL_4_WIKI_ALL)
        languageSelectionFragment.show(supportFragmentManager)
    }

    private fun setupLanguageSelectorMenuItem(menu: Menu) {
        val item = menu.findItem(R.id.menu_lang_selector)
        item.isVisible = true
        val rootView = item.actionView ?: return
        val selectedLang = rootView.findViewById<TextView>(R.id.txtSelectedLanguage)
        selectedLang.text = languageCode?.toUpperCase(Locale.ROOT) ?: ""
        rootView.setOnClickListener { loadLanguages() }
    }

    override fun onPrepareOptionsMenu(menu: Menu): Boolean {
        setupLanguageSelectorMenuItem(menu)
        return super.onPrepareOptionsMenu(menu)
    }

    override fun onBackPressed() {
        callBackPress()
    }

    private fun callBackPress() {
        if (binding.layoutList.visibility == View.VISIBLE || binding.root.findViewById<View>(R.id.layoutEmpty).visibility == View.VISIBLE) {
            binding.layoutEdit.makeVisible()
            binding.layoutList.makeGone()
            binding.root.findViewById<View>(R.id.layoutEmpty).makeGone()
        } else if (binding.layoutEdit.visibility == View.VISIBLE) {
            if (!TextUtils.isEmpty(binding.editFile.text)) {
                this.showConfirmBackDialog {
                    binding.layoutSelect.makeVisible()
                    binding.layoutEdit.makeGone()
                }
            } else {
                binding.layoutSelect.makeVisible()
                binding.layoutEdit.makeGone()
            }
        } else super.onBackPressed()
    }

    private fun callShowCaseUI() {
        if (!isFinishing && !isDestroyed) {
            if (isNotShowed(ShowCasePref.LIST_ITEM_SPELL_4_WIKI) && binding.recyclerView.visibility == View.VISIBLE && binding.recyclerView.getChildAt(0) != null) {
                val sequence = MaterialTapTargetSequence().setSequenceCompleteListener { showed(ShowCasePref.LIST_ITEM_SPELL_4_WIKI) }
                sequence.addPrompt(
                    getPromptBuilder(this@Spell4WordListActivity)
                        .setTarget(binding.recyclerView.getChildAt(0))
                        .setPrimaryText(R.string.sc_t_spell4wiki_list_item)
                        .setSecondaryText(R.string.sc_d_spell4wiki_list_item)
                )
                sequence.show()
            }
        }
    }
}